dynamic_instrumentation/src/point/mod.rs

```diff
@@ -0,0 +1,75 @@
+pub mod apply;
+pub mod build;
+
+use c2rust_analysis_rt::mir_loc::{self, EventMetadata};
+use rustc_middle::{
+    mir::{Body, HasLocalDecls, LocalDecls, Location, Place, Rvalue},
+    ty::TyCtxt,
+};
+use rustc_span::def_id::DefId;
+
+use crate::{arg::InstrumentationArg, hooks::Hooks, util::Convert};
+
+pub struct InstrumentationPoint<'tcx> {
+    id: usize,
+    pub loc: Location,
+    pub func: DefId,
+    pub args: Vec<InstrumentationArg<'tcx>>,
+    pub is_cleanup: bool,
+    pub after_call: bool,
+    pub metadata: EventMetadata,
+}
+
+pub struct InstrumentationAdder<'a, 'tcx: 'a> {
+    hooks: Hooks<'tcx>,
+    body: &'a Body<'tcx>,
+    instrumentation_points: Vec<InstrumentationPoint<'tcx>>,
+    assignment: Option<(Place<'tcx>, Rvalue<'tcx>)>,
+}
+
+impl<'a, 'tcx: 'a> InstrumentationAdder<'a, 'tcx> {
+    pub fn new(hooks: Hooks<'tcx>, body: &'a Body<'tcx>) -> Self {
+        Self {
+            hooks,
+            body,
+            instrumentation_points: Default::default(),
+            assignment: Default::default(),
+        }
+    }
+
+    pub fn tcx(&self) -> TyCtxt<'tcx> {
+        self.hooks.tcx()
```

Comment:
why does `hooks` need a `tcx`?

Reply:
It needs it to look through module items for the hooks.  I wasn't sure if I should put the `TyCtxt` in `Hooks`, but I tried other places and it didn't make as much sense or led to a bunch more code.

Reply:
i don't think `InstrumentationAdder` should rely on the `tcx` that `hooks` has. If we didn't have hooks, we'd still want `InstrumentationAdder` to have access to a `tcx`. In other words I think it should just get its own.

Reply:
I can do that.  That makes the struct bigger to store duplicate `TyCtxt`s, but it shouldn't matter much.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/560/commits/2d2813d49195fd8b378b4eeda6abb057b1b96ef9.